rpcperms/middleware_handler.go

```diff
@@ -382,24 +403,36 @@ func NewMessageInterceptionRequest(ctx context.Context,
 		return nil, err
 	}
 
-	rpcReq, ok := m.(proto.Message)
-	if !ok {
-		return nil, fmt.Errorf("msg is not proto message: %v", m)
+	req := &InterceptionRequest{
+		Type:        authType,
+		StreamRPC:   isStream,
+		Macaroon:    mac,
+		RawMacaroon: rawMacaroon,
+		FullURI:     fullMethod,
 	}
-	rawRequest, err := proto.Marshal(rpcReq)
-	if err != nil {
-		return nil, fmt.Errorf("cannot marshal proto msg: %v", err)
+
+	// The message is either a proto message or an error, we don't support
+	// any other types being intercepted.
+	switch t := m.(type) {
+	case proto.Message:
+		req.ProtoSerialized, err = proto.Marshal(t)
+		if err != nil {
+			return nil, fmt.Errorf("cannot marshal proto msg: %v",
+				err)
+		}
+		req.ProtoTypeName = string(proto.MessageName(t))
```

Comment:
Related to the comment above: I think if a user attempted to pass through an _actual_ proto error here, then we'd be able to detect that w/ `status.FromProto` (returns `nil` if it isn't an error) and then set `IsError` accordingly.